week-3/Voting/Multi/MultiRoundVoting.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.17;
+
+contract MultiRoundVoting {
+    // Voting related states
+    address public admin;
+    uint256 public votingRounds = 0;
+    uint256 public currentRound = 0;
+    bool public isRunning = false;
+    
+    // Result related states
+    address public winner = address(0);
+    uint256 public leadCount = 0;
+
+    // struct to hold voter details
+    struct Voter {
+        address voterId;
+        bool isVoted;
+        uint256 voteCount;
+    }
+
+    mapping(uint256 => mapping(address => Voter)) public voterList;
+
+    // events
+    event declareWinner(address indexed winner, uint256 indexed leadCount);
+
+    // Store initial values
+    constructor(uint256 _votingRounds) {
+        admin = msg.sender;
+        votingRounds = _votingRounds;
+    }
+
+    // register voter for voting
+    function registerVoter() public voterChecks roundRunningChecks {
+        require(voterList[currentRound+1][msg.sender].voterId == address(0), "Already registered");
+
+        // Register new voter in voting list with default values
+        voterList[currentRound+1][msg.sender] = Voter(msg.sender, false, 0);
+
+        return;
+    }
+
+    // vote for candidate
+    function vote (address _candidate) public voterChecks roundStopChecks {
+        // Make sure Voter is valid
+        require(voterList[currentRound][msg.sender].voterId != address(0), "Voter not registered!"); 
+        require(voterList[currentRound][msg.sender].isVoted == false, "Already voted!"); 
+
+        Voter storage candidate = voterList[currentRound][_candidate];
+
+        // Update voting status for voter
+        candidate.voteCount += 1; 
+        voterList[currentRound][msg.sender].isVoted = true; 
+
+        // Check if eligible for winner
+        if(leadCount < candidate.voteCount && candidate.voterId != winner){
```

Comment:
If the leading candidate gets vote.. the lead count is not increasing in this case.